src/Framework/FileClassifier.cs

```diff
@@ -0,0 +1,250 @@
+                }
+
+                // TODO: Once BuildEnvironmentHelper makes it from Shared into Framework, rework the code bellow. Hint: implement GetVsRootFromMSBuildAssembly() in BuildEnvironmentHelper
+
+                // Seems like MSBuild did not run from VS but from CLI.
+                // Identify current process and run it
+                string processName = Process.GetCurrentProcess().MainModule.FileName;
+                string processFileName = Path.GetFileNameWithoutExtension(processName);
+
+                if (string.IsNullOrEmpty(processFileName))
+                {
+                    return null;
+                }
+
+                string[] msBuildProcess = { "MSBUILD", "MSBUILDTASKHOST" };
+                if (msBuildProcess.Any(s =>
+                    processFileName.Equals(s, StringComparison.OrdinalIgnoreCase)))
+                {
+                    // Check if we're in a VS installation
+                    if (Regex.IsMatch(processName, $@".*\\MSBuild\\Current\\Bin\\.*MSBuild(?:TaskHost)?\.exe", RegexOptions.IgnoreCase))
+                    {
+                        return GetVsRootFromMSBuildAssembly(processName);
+                    }
+                }
+
+                return null;
+
+                static string GetVsRootFromMSBuildAssembly(string msBuildAssembly)
+                {
+                    return GetFolderAbove(msBuildAssembly,
+                        Path.GetDirectoryName(msBuildAssembly)?.EndsWith(@"\amd64", StringComparison.OrdinalIgnoreCase) == true
+                            ? 5
+                            : 4);
+                }
+
+                static string GetFolderAbove(string path, int count = 1)
+                {
+                    if (count < 1)
+                        return path;
+
+                    var parent = Directory.GetParent(path);
+
+                    while (count > 1 && parent?.Parent != null)
+                    {
+                        parent = parent.Parent;
+                        count--;
+                    }
+
+                    return parent?.FullName ?? path;
+                }
+            }
+#endif
+        }
+
+        /// <summary>
+        ///     Shared singleton instance.
+        /// </summary>
+        public static FileClassifier Shared => s_sharedInstance.Value;
+
+        /// <summary>
+        ///     Try add paths found into set of known immutable paths.
+        ///     Project files under any of these folders are considered non-modifiable.
+        /// </summary>
+        /// <remarks>
+        ///     This value is used by <see cref="IsNonModifiable" />.
+        ///     Files in the NuGet package cache are not expected to change over time, once they are created.
+        /// </remarks>
+        /// <remarks>
+        ///     Example value: <c>"C:\Users\myusername\.nuget\;D:\LocalNuGetCache\"</c>
+        /// </remarks>
+        public void RegisterImmutableDirectories(string? nuGetPackageFolders)
```

Comment:
I though about that too, but rejected it.
Only the calling side knows if the value shall be trimmed as in case of MSBuild properties which, being human editable, are prone to either end spaces.
I actually do not think it matter as much. I am resolving it, if you feel string about it, feel free to unresolve it please.